Overlap computations are broken with rotate, translate, scale properties

graouts opened the pull request:
#### 9db2da0e4d0a921511e1c740351e3bba0f400e7a
<pre>
Overlap computations are broken with rotate, translate, scale properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=236019">https://bugs.webkit.org/show_bug.cgi?id=236019</a>

Reviewed by Simon Fraser.

We extend the extent computation for transform animations to also include the set
of individual transform properties: `translate`, `scale` and `rotate`.

Previously, the function used to compute that extent, `computeExtentOfTransformAnimation()`,
had two separate code paths: one for `matrix()` functions and another for the other function
types. Now we use a single approach which, for each keyframe:

1. computes the blended style for that keyframe using `setAnimatedPropertiesInStyle()`,
2. computes the transform inclusive of all properties using `RenderStyle::applyTransform()`.

This has the benefit of using the exact same values that will be hit during the animation.
Thanks to this approach, we can compute slightly tighter extents in the existing tests.

We also add new tests for individual transform properties.

Tests: compositing/layer-creation/multiple-keyframes-individual-transform-properties-animation-overlap.html
       compositing/layer-creation/scale-rotation-individual-properties-animation-overlap.html
       compositing/layer-creation/translate-scale-individual-properties-animation-overlap.html

* LayoutTests/compositing/layer-creation/multiple-keyframes-animation-overlap-expected.txt:
* LayoutTests/compositing/layer-creation/multiple-keyframes-individual-transform-properties-animation-overlap-expected.txt: Added.
* LayoutTests/compositing/layer-creation/multiple-keyframes-individual-transform-properties-animation-overlap.html: Added.
* LayoutTests/compositing/layer-creation/scale-rotation-animation-overlap-expected.txt:
* LayoutTests/compositing/layer-creation/scale-rotation-animation-overlap.html:
* LayoutTests/compositing/layer-creation/scale-rotation-individual-properties-animation-overlap-expected.txt: Added.
* LayoutTests/compositing/layer-creation/scale-rotation-individual-properties-animation-overlap.html: Added.
* LayoutTests/compositing/layer-creation/translate-scale-animation-overlap-expected.txt:
* LayoutTests/compositing/layer-creation/translate-scale-individual-properties-animation-overlap-expected.txt: Added.
* LayoutTests/compositing/layer-creation/translate-scale-individual-properties-animation-overlap.html: Added.
* LayoutTests/platform/ios/compositing/layer-creation/scale-rotation-animation-overlap-expected.txt:
* LayoutTests/platform/ios/compositing/layer-creation/scale-rotation-individual-properties-animation-overlap-expected.txt: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::isRunningAcceleratedTransformRelatedAnimation const):
(WebCore::KeyframeEffect::animatesMotionPath const):
(WebCore::KeyframeEffect::preventsAcceleration const):
(WebCore::KeyframeEffect::transformRelatedPropertyDidChange):
(WebCore::KeyframeEffect::computeExtentOfTransformAnimation const):
(WebCore::propertiesContainTransformRelatedProperty): Deleted.
(WebCore::KeyframeEffect::computeTransformedExtentViaTransformList const): Deleted.
(WebCore::KeyframeEffect::computeTransformedExtentViaMatrix const): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::animatablePropertiesContainTransformRelatedProperty):
* Source/WebCore/animation/WebAnimationUtilities.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::currentTransform const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::computeAnimationExtent const):
(WebCore::Styleable::isRunningAcceleratedTransformRelatedAnimation const):
* Source/WebCore/style/Styleable.h:

Canonical link: <a href="https://commits.webkit.org/303045@main">https://commits.webkit.org/303045@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea89137ac7fdf95a25462ca56c182e43c9aa2db6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ style](https://ews-build.webkit.org/#/builders/38/builds/131033 "Pass") | [✅ ios](https://ews-build.webkit.org/#/builders/159/builds/3358 "Pass") | [✅ mac](https://ews-build.webkit.org/#/builders/138/builds/41992 "Pass") | [✅ wpe](https://ews-build.webkit.org/#/builders/5/builds/138475 "Pass") | [✅ win](https://ews-build.webkit.org/#/builders/59/builds/82719 "Pass") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7aeb2e33-527b-4432-9d68-c6e46d1780ab) 
| [✅ bindings](https://ews-build.webkit.org/#/builders/9/builds/132904 "Pass") | [✅ ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3343 "Pass") | [✅ mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3199 "Pass") | [❌ wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99823 "Found 1 new test failure: fast/repaint/translate-animation-repaint.html (failure)") | [  ~~win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5e1e4442-45c7-4de9-9e25-d4bc8378f927) 
| [✅ webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133979 "Pass") | [✅ ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2388 "Pass") | [✅ api-mac](https://ews-build.webkit.org/#/builders/18/builds/117323 "Pass") | [✅ api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80532 "Pass") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d9fe139e-f4e1-4a61-a703-278143f809e2) 
| | [✅ ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2309 "Pass") | [⏳ api-mac-debug](https://ews-build.webkit.org/#/builders/138/builds/41992 "Waiting to run tests") | [✅ wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81722 "Pass") | | 
| | [✅ api-ios](https://ews-build.webkit.org/#/builders/13/builds/110916 "Pass") | [  ~~mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/35448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ gtk](https://ews-build.webkit.org/#/builders/2/builds/140969 "Pass") | | 
| | [✅ vision](https://ews-build.webkit.org/#/builders/153/builds/3100 "Pass") | [✅ mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35943 "Pass") | [❌ gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108339 "Found 1 new test failure: fast/repaint/translate-animation-repaint.html (failure)") | | 
| | [✅ vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3146 "Pass") | [✅ mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2775 "Pass") | [✅ api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108296 "Pass") | | 
| | [✅ vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2348 "Pass") | [✅ mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113639 "Pass") | [✅ playstation](https://ews-build.webkit.org/#/builders/134/builds/56153 "Pass") | | 
| [✅ unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20396 "Pass") | [✅ tv](https://ews-build.webkit.org/#/builders/158/builds/3168 "Pass") | [✅ mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32062 "Pass") | | | 
| | [✅ tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2989 "Pass") | [✅ mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66563 "Pass") | | | 
| | [✅ watch](https://ews-build.webkit.org/#/builders/163/builds/3189 "Pass") | | | | 
| | [✅ watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3098 "Pass") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -1522,8 +1522,12 @@ void RenderLayerBacking::updateGeometry(const RenderLayer* compositedAncestor)
-    if (auto styleable = Styleable::fromRenderer(renderer()))
-        isRunningAcceleratedTransformAnimation = styleable->isRunningAcceleratedAnimationOfProperty(CSSPropertyTransform);
+    if (auto styleable = Styleable::fromRenderer(renderer())) {
+        isRunningAcceleratedTransformAnimation = styleable->isRunningAcceleratedAnimationOfProperty(CSSPropertyTransform)
```
I wonder if calling `isRunningAcceleratedAnimationOfProperty()` 4 times here has enough overhead to justify making a single function that does it.

graouts replied:
I'll make that change, if anything it'll read better and abstract away which properties are concerned (this will be expanded with threaded animations as we add support for animating CSS Motion Path properties).

graouts replied:
This was solved by the addition of `Styleable::isRunningAcceleratedTransformRelatedAnimation()`.

smfr commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -2483,121 +2479,84 @@ const RenderStyle& KeyframeEffect::currentStyle() const
-            else
-                continue;
-        }
+        auto blendedStyleForKeyframe = RenderStyle::clonePtr(*unanimatedStyle);
```
Let's keep an eye on the perf impact.

graouts committed:
Overlap computations are broken with rotate, translate, scale properties
https://bugs.webkit.org/show_bug.cgi?id=236019

Reviewed by Simon Fraser.

We extend the extent computation for transform animations to also include the set
of individual transform properties: `translate`, `scale` and `rotate`.

Previously, the function used to compute that extent, `computeExtentOfTransformAnimation()`,
had two separate code paths: one for `matrix()` functions and another for the other function
types. Now we use a single approach which, for each keyframe:

1. computes the blended style for that keyframe using `setAnimatedPropertiesInStyle()`,
2. computes the transform inclusive of all properties using `RenderStyle::applyTransform()`.

This has the benefit of using the exact same values that will be hit during the animation.
Thanks to this approach, we can compute slightly tighter extents in the existing tests.

We also add new tests for individual transform properties.

Tests: compositing/layer-creation/multiple-keyframes-individual-transform-properties-animation-overlap.html
       compositing/layer-creation/scale-rotation-individual-properties-animation-overlap.html
       compositing/layer-creation/translate-scale-individual-properties-animation-overlap.html

* LayoutTests/compositing/layer-creation/multiple-keyframes-animation-overlap-expected.txt:
* LayoutTests/compositing/layer-creation/multiple-keyframes-individual-transform-properties-animation-overlap-expected.txt: Added.
* LayoutTests/compositing/layer-creation/multiple-keyframes-individual-transform-properties-animation-overlap.html: Added.
* LayoutTests/compositing/layer-creation/scale-rotation-animation-overlap-expected.txt:
* LayoutTests/compositing/layer-creation/scale-rotation-animation-overlap.html:
* LayoutTests/compositing/layer-creation/scale-rotation-individual-properties-animation-overlap-expected.txt: Added.
* LayoutTests/compositing/layer-creation/scale-rotation-individual-properties-animation-overlap.html: Added.
* LayoutTests/compositing/layer-creation/translate-scale-animation-overlap-expected.txt:
* LayoutTests/compositing/layer-creation/translate-scale-individual-properties-animation-overlap-expected.txt: Added.
* LayoutTests/compositing/layer-creation/translate-scale-individual-properties-animation-overlap.html: Added.
* LayoutTests/platform/ios/compositing/layer-creation/scale-rotation-animation-overlap-expected.txt:
* LayoutTests/platform/ios/compositing/layer-creation/scale-rotation-individual-properties-animation-overlap-expected.txt: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::isRunningAcceleratedTransformRelatedAnimation const):
(WebCore::KeyframeEffect::animatesMotionPath const):
(WebCore::KeyframeEffect::preventsAcceleration const):
(WebCore::KeyframeEffect::transformRelatedPropertyDidChange):
(WebCore::KeyframeEffect::computeExtentOfTransformAnimation const):
(WebCore::propertiesContainTransformRelatedProperty): Deleted.
(WebCore::KeyframeEffect::computeTransformedExtentViaTransformList const): Deleted.
(WebCore::KeyframeEffect::computeTransformedExtentViaMatrix const): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::animatablePropertiesContainTransformRelatedProperty):
* Source/WebCore/animation/WebAnimationUtilities.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::currentTransform const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::computeAnimationExtent const):
(WebCore::Styleable::isRunningAcceleratedTransformRelatedAnimation const):
* Source/WebCore/style/Styleable.h:

Canonical link: https://commits.webkit.org/303045@main
